### PR TITLE
requirement: Only copy FIFO files into temporary files

### LINKED ISSUE #552: Could not open requirements file: No such file or directory on Linux
## Bug description

When `requirements.txt` file has another requirements file referenced inside it with `-r child/component/requirements.txt`, `pip-audit -r requirements.txt` fails.

## Reproduction steps

1. Install pip-audit v2.5.0
2. Create a requirements file for child component at `child/component/requirements.txt` with any dependency.
3. Create a requirements file at project root `requirements.txt`
4. Reference `-r child/component/requirements.txt` inside the `requirements.txt` at root level.
5. Run `pip-audit -r requirements.txt`
6. pip-audit fails:
   ```
   ERROR:pip_audit._virtual_env:internal pip failure: ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/tmp/child/component/requirements.txt'
   ```

## Expected behavior

No error.

## Screenshots and logs

Log:
```
DEBUG:pip_audit._cli:parsed arguments: Namespace(local=False, requirements=[<_io.TextIOWrapper name='requirements.txt' mode='r' encoding='UTF-8'>], project_path=None, format=<OutputFormatChoice.Columns: 'columns'>, vulnerability_service=<VulnerabilityServiceChoice.Pypi: 'pypi'>, dry_run=False, strict=False, desc=<VulnerabilityDescriptionChoice.Auto: 'auto'>, cache_dir=None, progress_spinner=<ProgressSpinnerChoice.On: 'on'>, timeout=15, paths=[], verbose=1, fix=False, require_hashes=False, index_url='https://pypi.org/simple/', extra_index_urls=[], skip_editable=False, no_deps=False, output=PosixPath('stdout'), ignore_vulns=[])
ERROR:pip_audit._virtual_env:internal pip failure: ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/tmp/child/component/requirements.txt'

ERROR:pip_audit._cli:Failed to install packages: ['/tmp/tmph391p24m/bin/python', '-m', 'pip', 'install', '--index-url', 'https://pypi.org/simple/', '--dry-run', '--report', '/tmp/tmpxk8qhmm1', '-r', '/tmp/tmp8vdjs_9y']
```

child/component/requirements.txt
```
mypy-boto3-dynamodb >= 1.26.0
```

requirements.txt
```
-r child/component/requirements.txt
```

## Platform information

* OS name and version: 
  Operating System: Amazon Linux 2
  Kernel: Linux 5.4.228-141.415.amzn2int.x86_64
  Architecture: x86-64
* `pip-audit` version (`pip-audit -V`): 
   pip-audit 2.5.0
* Python version (`python -V` or `python3 -V`): 
   Python 3.9.13
* `pip` version (`pip -V` or `pip3 -V`): 
   pip 22.0.4

## Additional context

This started happening after `2.5.0` release on March 16th.  With previous versions there was no error.

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
Closes #552

### CODE REVIEW

**Comment by woodruffw** [comment]
> I still need to do something to get test coverage for FIFO files. I'm thinking of creating one with `os.mkfifo` and then passing it into the `RequirementSource`.

That SGTM!

**Review by woodruffw — APPROVED** [review]
LGTM but let's get that unit test in before merging.

File: CHANGELOG.md [hunk 1]
```diff
@@ -26,6 +26,9 @@ All versions prior to 0.0.9 are untracked.
 * Fixed a crash caused by invalid UTF-8 sequences in subprocess outputs
   ([#572](https://github.com/pypa/pip-audit/pull/572))
 
+* Fixed a crash caused by auditing requirements files that refer to other
+  requirements files ([#568](https://github.com/pypa/pip-audit/pull/568))
```
[the remark this answers is not in the document]
Comment (tetsuo-cpp) [reply CHANGELOG.md:30]: Whoops, not sure why I did that lol.

[the remark this answers is not in the document]
Comment (tetsuo-cpp) [reply CHANGELOG.md]: [41e7f41](https://github.com/pypa/pip-audit/pull/568/commits/41e7f41b30f35c9b597f3250ebcda773233bde52)

File: pip_audit/_dependency_source/requirement.py [hunk 2]
```diff
@@ -81,42 +81,47 @@ def collect(self) -> Iterator[Dependency]:
...
+            # Now pass the list of filenames into the rest of our logic.
+            yield from self._collect_from_files(collect_files)
         finally:
-            # Since we disabled automatically deletion for these temporary files, we need to
-            # manually delete them on the way out.
+            # Since we disabled automatically deletion for these temporary
+            # files, we need to manually delete them on the way out.
             for t in tmp_files:
-                os.unlink(t)
+                os.unlink(str(t))
```
[the remark this answers is not in the document]
Comment (tetsuo-cpp) [reply pip_audit/_dependency_source/requirement.py]: [b9b7f86](https://github.com/pypa/pip-audit/pull/568/commits/b9b7f866c2c54dd8d5b02ddb5a2be57c16f689fb)

**Pull request merged**